src/mlpack/methods/ann/gan/gan.hpp

```diff
@@ -0,0 +1,357 @@
+ * in unsupervised machine learning, implemented by a system of two neural
+ * networks contesting with each other in a zero-sum game framework. This
+ * technique can generate photographs that look at least superficially
+ * authentic to human observers, having many realistic characteristics.
+ *
+ * For more information, see the following paper:
+ *
+ * @code
+ * @article{Goodfellow14,
+ *   author    = {Ian J. Goodfellow, Jean Pouget-Abadi, Mehdi Mirza, Bing Xu,
+ *                David Warde-Farley, Sherjil Ozair, Aaron Courville and
+ *                Yoshua Bengio},
+ *   title     = {Generative Adversarial Nets},
+ *   year      = {2014},
+ *   url       = {http://arxiv.org/abs/1406.2661},
+ *   eprint    = {1406.2661},
+ * }
+ * @endcode
+ *
+ * @tparam Model The class type of Generator and Discriminator.
+ * @tparam InitializationRuleType Type of Initializer.
+ * @tparam Noise The noise function to use.
+ * @tparam PolicyType The GAN variant to be used (GAN, DCGAN, WGAN or WGANGP).
+ */
+template<
+  typename Model,
+  typename InitializationRuleType,
+  typename Noise,
+  typename PolicyType = StandardGAN
+>
+class GAN
+{
+ public:
+  /**
+   * Constructor for GAN class.
+   *
+   * @param trainData The real data.
+   * @param generator Generator network.
+   * @param discriminator Discriminator network.
+   * @param batchSize Batch size to be used for training.
+   * @param generatorUpdateStep Number of steps to train Discriminator
+   *                            before updating Generator.
+   * @param preTrainSize Number of pre-training steps of Discriminator.
+   * @param multiplier Ratio of learning rate of Discriminator to the Generator.
+   * @param clippingParameter Weight range for enforcing Lipschitz constraint.
+   * @param lambda Parameter for setting the gradient penalty.
+   */
+  GAN(arma::mat& trainData,
+      Model& generator,
+      Model& discriminator,
+      InitializationRuleType initializeRule,
+      Noise noiseFunction,
+      size_t noiseDim,
+      size_t batchSize,
+      size_t generatorUpdateStep,
+      size_t preTrainSize,
+      double multiplier,
+      double clippingParameter = 0.01,
+      double lambda = 10.0);
+
+  // Reset function.
+  void Reset();
+
+  // Train function.
+  template<typename OptimizerType>
+  void Train(OptimizerType& Optimizer);
+
+  /**
+   * Evaluate function for the Standard GAN and DCGAN.
+   * This function gives the performance of the Standard GAN or DCGAN on the
+   * current input.
+   *
+   * @param parameters The parameters of the network.
+   * @param i Index of the current input.
+   * @param batchSize Variable to store the present number of inputs.
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, StandardGAN>::value ||
+                          std::is_same<Policy, DCGAN>::value, double>::type
+  Evaluate(const arma::mat& parameters,
+           const size_t i,
+           const size_t batchSize);
```

Comment:
The various `enable_if`s make me wonder if it's better to put the `Evaluate()` function in the `Policy` class, and have this just be a wrapper to call out to `Policy::Evaluate()`.  I know it is very late in the design process so maybe it is not worth revisiting, and I think this can be fine, but maybe it is an opportunity for improvement later or something.

Reply:
Actually, we started the implementation by doing that, but unfortunately the templatized classes were unable to access the `GAN` class' internal members. So we chose this approach instead.

Reply:
Right, another idea would be to make the templatized classes accept all the internal `GAN` class members that would be needed in their signatures.  I think there would be no issue with that, but it is up to you.  I'm not sure if it would be uglier or prettier.

Reply:
That's exactly the next approach we went to. But @zoq and I ended up agreeing that since `Evaluate()` and `Gradient()` tend to access all the members, it would look a lot worse and using it would be even more frustrating.

Reply:
Right, I guess you have already had this discussion then.  Sorry to make you write it down again :)